using_restricted_array.rb

```diff
@@ -7,41 +7,236 @@
+=begin
+  1. We know length of array
+  2. it contains integets
+  3. it is unsorted
+
+  Want:
+  1. find max value in array
+  2. once found return max
+
+
+  Algorithm
+  1. set i to 0
+  2. set max to value at i
+  3. while i < length
+    - if value at i > max
+      -  replace max with value at i
+    - increment i by 1
+  4.  return max
+=end
+
+  i = 0
+  max = array[0]
+
+  #
+  # while i < length(array)
+  #   if array[i] > max
+  #     max = array[i]
+  #   end
+  #   i += 1
+  # end
+  # return max
+
+  # Another way
+  # (1..length).times do |value|
+  #   if value > max
+  #     max = array[value]
+  #   end
+  #   # i += 1
+  # end
+  # return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+=begin
+    1. We know length of array
+    2. it contains integets
+    3. it is unsorted
+
+    Want:
+    1. find min value in array
+    2. once found return min
+
+
+    Algorithm
+    1. set i to 0
+    2. set min to value at i
+    3. while i < length
+      - if value at i < min
+        -  replace min with value at i
+        - increment i by 1
+    4.  return min
+=end
+
+  i = 0
+  min = array[0]
+
+  while i < length
+    if array[i] < min
+      min = array[i]
+    end
+    i += 1
+  end
+  return min
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+=begin
+      1. We know length of array
+      2. it contains integets
+      3. it is unsorted
+
+      Want:
+      1. reverse array in place
+
+      Algorithm
+      1. check contents, none then done (this step is for me to remember in future.)
+      2. set i to 0
+      3. set j to lemgth - 1
+      3. while i < j
+        - swap values at i and j
+        - increment i by 1
+        - decrement j by 1
+=end
+  i = 0
+  j = length-1
+
+  while i < j
+    array[i] = array[i] + array[j]
+    array[j] = array[i] - array[j]
+    array[i] = array[i] - array[j]
+
+    i += 1
+    j -= 1
+
+  end
+
+
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+=begin
+        1. We know length of array
+        2. it contains integets
+        3. it is unsorted
+
+        Want:
+        1. find value with binary search
+        2. Returns true if found
+        3. false otherwise
+
+        Algorithm
+        1. check contents, none then done (this step is for me to remember in future.)
+        2. set low to first index
+        3. set high to last index
+        3. while low < high
+          - index to mid
+          - if match, return true
+          - if value at index mid is higher, eliminate half and set high to mid
+          - if value at index mid is lower, eliminate half and set low to mid.
+
+=end
+
+  low = 0
+  high = length-1
+
```

Comment:
Please take a look at the updated binary_search slide. 

Here's the considerations I was missing in the algorithm on the slide before the update: (1) what if there is only one element in the array and it is the value_to_find? In this case, it used to return false. (To fix this, add an additional if statement after the loop ends, to check if array[low] is the value_to_find.) (2) Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, min = 0, max=1, mid=0. Value is not at index 0. So, min = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. (To fix this, update max to be mid -1, if value_to_find is less than mid. And update min = mid+1, if value_to_find is greater than mid.)